torch/testing/_asserts.py

```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
+    if exc:
+        return exc, None
 
+    return None, tensor
 
-def _parse_sequences(actual: Sequence, expected: Sequence) -> Tuple[Optional[Exception], Optional[_ParsedInputs]]:
-    """Parses sequences of scalar-or-array-like pairs.
 
-    Regardless of the input types, the sequences are returned as :class:`list`.
+def _to_tensor_pair(actual: Any, expected: Any) -> Tuple[Optional[Exception], Optional[_TensorPair]]:
+    """Converts a scalar-or-array-like pair to a :class:`_TensorPair`.
 
     Args:
-        actual: Actual sequence array-or-scalar-likes.
-        expected: Expected sequence array-or-scalar-likes.
+        actual (Any): Actual array-or-scalar-like.
+        expected (Any): Expected array-or-scalar-like.
 
     Returns:
-        (Optional[Exception], Optional[_ParsedInputs]): The two elements are orthogonal, i.e. if the first ``is None``
-            the second will not and vice versa. Returns a :class:`AssertionError` if the length of :attr:`actual` and
-            :attr:`expected` does not match. Additionally, returns any exception from
-            :func:`_parse_array_or_scalar_like_pair`.
+        (Tuple[Optional[Exception], Optional[_TensorPair]]): The two elements are orthogonal, i.e. if the first
+            ``is None`` the second will be valid and vice versa. Returns a :class:`UsageError` if :attr:`actual` and
+            :attr:`expected` do not have the same type. Additionally, returns any exception from :func:`_to_tensor`.
     """
     exc: Optional[Exception]
 
-    actual_len = len(actual)
-    expected_len = len(expected)
-    if actual_len != expected_len:
-        exc = AssertionError(f"The length of the sequences mismatch: {actual_len} != {expected_len}")
+    if type(actual) is not type(expected):
+        exc = UsageError(
+            f"Apart from a containers, type equality is required, but got {type(actual)} and {type(expected)} instead."
+        )
         return exc, None
 
-    actual_lst = []
-    expected_lst = []
-    for idx in range(actual_len):
-        exc, result = _parse_array_or_scalar_like_pair(actual[idx], expected[idx])
-        if exc:
-            exc = _amend_error_message(exc, f"{{}}\n\n{_SEQUENCE_MSG_FMTSTR.format(idx)}")
-            return exc, None
+    exc, actual = _to_tensor(actual)
+    if exc:
+        return exc, None
 
-        result = cast(_ParsedInputs, result)
-        actual_lst.append(cast(Tensor, result.actual))
-        expected_lst.append(cast(Tensor, result.expected))
+    exc, expected = _to_tensor(expected)
+    if exc:
+        return exc, None
 
-    return None, _ParsedInputs(actual_lst, expected_lst)
+    return None, _TensorPair(actual, expected)
 
 
-def _parse_mappings(actual: Mapping, expected: Mapping) -> Tuple[Optional[Exception], Optional[_ParsedInputs]]:
-    """Parses sequences of scalar-or-array-like pairs.
+def _parse_inputs(actual: Any, expected: Any) -> Tuple[Optional[Exception], Optional[Union[_TensorPair, List, Dict]]]:
```

Comment:
Before we returned two corresponding tensors or tensor containers. Now this function bundles the actual and expected tensors together into one `_TensorPair`. I plan to use this for all internal functions that happen "after" the input parsing.

Reply:
What's the benefit of pairing them in an object vs. just passing actual and expected directly?

Reply:
Especially when having nested containers, it is harder two maintain two inputs with 1-to-1 correspondence instead of one where each item is a pair of two tensors. This function performs this transformation.